### ISSUE TITLE
Failed validation

### ISSUE THREAD

**Issue by erikd** [input-output-hk/cardano-db-sync#60]
Since adding a new validation to `cardano-db-tool` a couple of days ago, I have been running validation on my `db-sync` node periodically. I have now seen the first failure:
```
nix@nix:~/cardano-db-sync$ PGPASSFILE=config/pgpass ./cardano-db-tool validate 
Total supply plus fees at block 1515226 is same as genesis supply: ok
Validate total supply decreasing from block 1515226 to block 3770668: ok
Epoch table entries for epochs [0..184] are correct: ok
All txs for blocks in epoch 185 are present: Failed on block no 4010399: expected tx count of 0 but got 1
All txs for blocks in epoch 169 are present: ok
```
Since the validation always run on the most recent epoch (currently 185) repeated runs always fail the same way, so this is definitely an error, but there are no errors in the logs that I can see.

Even more strangely this SQL query gives more than one row with `block_no = 4010399`:
```
cexplorer=# select id, slot_no, block_no, previous, slot_leader, time, tx_count from block
            where block_no = 4010399 ; 
   id    | slot_no | block_no | previous | slot_leader |        time         | tx_count 
---------+---------+----------+----------+-------------+---------------------+----------
 4011986 | 4012493 |  4010399 |  4011984 |           7 | 2020-04-09 17:22:31 |        0
 4011985 | 4012492 |  4010399 |  4011984 |           6 | 2020-04-09 17:22:11 |        1
```
Looking for blocks which have either of those as a previous:
```
cexplorer=# select id, slot_no, block_no, previous, slot_leader, time, tx_count from block
                 where previous = 4011986 ; 
   id    | slot_no | block_no | previous | slot_leader |        time         | tx_count 
---------+---------+----------+----------+-------------+---------------------+----------
 4011987 | 4012494 |  4010400 |  4011986 |           8 | 2020-04-09 17:22:51 |        0
(1 row)

cexplorer=# select id, slot_no, block_no, previous, slot_leader, time, tx_count from block
                   where previous = 4011985 ; 
 id | slot_no | block_no | previous | slot_leader | time | tx_count 
----+---------+----------+----------+-------------+------+----------
(0 rows)
```
This suggests that a rollback happened and indeed, in the logs I see:
```
[db-sync-node:Info:55] [2020-04-09 17:22:13.56 UTC] insertABlock: slot 4012492, block 4010399, hash ee8b35b3ea713f57c5000b942178cef876d17a90be44a1830f9b2f0e187b9914
[db-sync-node:Info:55] [2020-04-09 17:22:54.90 UTC] No rollback required: chain tip slot is 4012491
[db-sync-node:Info:55] [2020-04-09 17:22:54.91 UTC] insertABlock: slot 4012493, block 4010399, hash c928c2e82a3efbf1e3f0382636e4eac0933510cdc51f65266b85c16084d8feb2
[db-sync-node:Info:55] [2020-04-09 17:22:55.10 UTC] epochPluginInsertBlock: Inserting row in epoch table for epoch 185
[db-sync-node:Info:55] [2020-04-09 17:22:55.11 UTC] insertABlock: slot 4012494, block 4010400, hash ba0be729a025373d23ce419a584711f158ca904d3b86eae26ad094422b33cd77
```
which suggests something like a restart of db sync thread.

I tried deleting the `Block` table entry that no subsequent block used as a previous block but foreign key constraints make that difficult.

**Comment by erikd** [comment]
It seems persistent is not able to correctly add `ON DELETE CASCADE` to foreign keys: https://github.com/yesodweb/persistent/issues/910

**Comment by erikd** [comment]
Ok, there seems to be a disagreement on where to rollback to.

The consensus layer specifies the rollback point to be:
```
data ByronBlock = ByronBlock {
      byronBlockRaw    :: !(CC.ABlockOrBoundary ByteString)
    , byronBlockSlotNo :: !SlotNo
    , byronBlockHash   :: !ByronHash
    }
  deriving (Eq, Show)
```
where as `db-sync` works on the block number. This has now been corrected in #61 which may fix this issue.

I have also added a "rollback to a specific slot number" command to `cardano-db-tool` and improved the debug output for the rollback decision logic.
